torch/_dynamo/variables/tensor.py

```diff
@@ -644,6 +644,46 @@ def redistribute_fn_with_prim_types(x):
                 ),
                 **options,
             )
+        elif name == "register_hook":
+            # see [On tensor.register_hook]
+            assert len(args) == 1
+            fn_var = args[0]
+
+            if isinstance(fn_var, variables.NestedUserFunctionVariable):
+                # NestedUserFunctionVariable don't carry their fn, but reconstruction builds it
+                # This should not be onerous to support when needed.
+                unimplemented("NYI - lambda variables as hooks")
+            elif isinstance(fn_var, variables.functions.FunctoolsPartialVariable):
+                fn = fn_var.as_python_constant()
+                name = fn_var.func.fn.__name__
+            else:
```

Comment:
so... what kind of variable is fn_var here? 🤔

Reply:
Any callable - of which we have many.

Reply:
yes but why is func guaranteed to exist? It's not right?

Reply:
Sorry, which side of the else are you talking about?

Reply:
err, fn, not func. Below here.

Reply:
Its an invariant that anything that gets to here has a func, but we can graph break if its a surprising type